File: core/src/main/java/ru/pocketbyte/locolaser/config/parser/BaseTableSourceConfigParser.java
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser.config.parser;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import ru.pocketbyte.locolaser.config.source.BaseTableSourceConfig;
import ru.pocketbyte.locolaser.config.source.SourceConfig;
import ru.pocketbyte.locolaser.config.source.SourceSetConfig;
import ru.pocketbyte.locolaser.exception.InvalidConfigException;
import ru.pocketbyte.locolaser.utils.JsonParseUtils;
import ru.pocketbyte.locolaser.utils.TextUtils;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * @author Denis Shurygin
 */
public abstract class BaseTableSourceConfigParser<SourceConfigClass extends BaseTableSourceConfig> implements SourceConfigParser<SourceConfig> {

    public static final String TYPE = "type";
    public static final String COLUMN_KEY = "column_key";
    public static final String COLUMN_QUANTITY = "column_quantity";
    public static final String COLUMN_LOCALES = "column_locales";
    public static final String COLUMN_COMMENT = "column_comment";

    /**
     * Creates a new Source config object by it's name.
     *
     * @param type Source type.
     * @return Source object.
     * @throws InvalidConfigException if source type is unknown.
     */
    protected abstract SourceConfigClass sourceByType(String type) throws InvalidConfigException;

    /**
     * Parse Source from JSON object.
     *
     * @param sourceObject JSON object that contain source config properties.
     * @return Parsed source object.
     * @throws InvalidConfigException
     */
    public SourceConfig parse(Object sourceObject, boolean throwIfWrongType) throws InvalidConfigException {
        if (sourceObject instanceof JSONObject) {
            return parseFromJson((JSONObject) sourceObject);
        } else if (sourceObject instanceof JSONArray) {
            SourceConfig defaultConfig = null;
            Set<SourceConfig> configs = new HashSet<>(((JSONArray) sourceObject).size());
            for (Object object : (JSONArray) sourceObject) {
                if (object instanceof JSONObject) {
                    SourceConfig config = parseFromJson((JSONObject) object);
                    configs.add(config);

                    if (defaultConfig == null)
                        defaultConfig = config;
                }
                else
                    throw new InvalidConfigException("Source array must contain JSONObjects.");
            }
            return new SourceSetConfig(configs, defaultConfig);
        } else if (throwIfWrongType){
            throw new InvalidConfigException("Source must be a JSONObject or JSONArray.");
        }
        return null;
    }

    protected SourceConfigClass parseFromJson(JSONObject configJson) throws InvalidConfigException {
        String type = JsonParseUtils.getString(configJson, TYPE, ConfigParser.SOURCE, false);
        SourceConfigClass source = sourceByType(type);
        fillFromJSON(source, configJson);
        validate(source);
        return source;
    }

    /**
     * Fill source object from JSON.
     *
     * @param source      Source to fill.
     * @param configJson JSON object that contain source config properties.
     * @throws InvalidConfigException if config has some logic errors or doesn't contain some required fields.
     */
    protected void fillFromJSON(SourceConfigClass source, JSONObject configJson) throws InvalidConfigException {
        source.setKeyColumn(JsonParseUtils.getString(configJson, COLUMN_KEY, ConfigParser.SOURCE, true));
        source.setQuantityColumn(JsonParseUtils.getString(configJson, COLUMN_QUANTITY, ConfigParser.SOURCE, false));
        source.setCommentColumn(JsonParseUtils.getString(configJson, COLUMN_COMMENT, ConfigParser.SOURCE, false));

        List<String> locales = JsonParseUtils.getStrings(configJson, COLUMN_LOCALES, ConfigParser.SOURCE, true);
        if (locales != null)
            source.setLocaleColumns(new HashSet<>(locales));
    }

    protected void validate(SourceConfigClass source) throws InvalidConfigException {
        if (TextUtils.isEmpty(source.getKeyColumn()))
            throw new InvalidConfigException("\"" + ConfigParser.SOURCE + "." + COLUMN_KEY + "\" is not set.");
        if (source.getLocaleColumns() == null)
            throw new InvalidConfigException("\"" + ConfigParser.SOURCE + "." + COLUMN_KEY + "\" is not set.");
        if (source.getLocaleColumns().size() == 0)
            throw new InvalidConfigException("\"" + ConfigParser.SOURCE + "." + COLUMN_LOCALES + "\" must contain at least one item.");
    }
}

File: core/src/main/java/ru/pocketbyte/locolaser/config/parser/BaseTableSourceConfigParser.kt
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser.config.parser

import org.json.simple.JSONObject
import ru.pocketbyte.locolaser.config.parser.ConfigParser.Companion.SOURCE
import ru.pocketbyte.locolaser.config.resources.ResourcesConfig
import ru.pocketbyte.locolaser.config.resources.BaseTableResourcesConfigBuilder
import ru.pocketbyte.locolaser.exception.InvalidConfigException
import ru.pocketbyte.locolaser.utils.json.JsonParseUtils

/**
 * @author Denis Shurygin
 */
@Deprecated("JSON configs is deprecated feature. You should use Gradle config configuration")
abstract class BaseTableSourceConfigParser<ConfigBuilder : BaseTableResourcesConfigBuilder<*>>
    : ResourcesConfigParser<ResourcesConfig> {

    companion object {
        const val TYPE = "type"
        const val COLUMN_KEY = "column_key"
        const val COLUMN_QUANTITY = "column_quantity"
        const val COLUMN_COMMENT = "column_comment"
        const val COLUMN_METADATA = "column_metadata"
    }

    /**
     * Creates a new config builder object by it's name.
     *
     * @param type Source type.
     * @return builder object.
     * @throws InvalidConfigException if type is unknown.
     */
    @Throws(InvalidConfigException::class)
    protected abstract fun builderByType(type: String?, throwIfWrongType: Boolean): ConfigBuilder?

    /**
     * Parse Source from JSON object.
     *
     * @param resourceObject JSON object that contain source config properties.
     * @return Parsed source object.
     * @throws InvalidConfigException
     */
    @Throws(InvalidConfigException::class)
    override fun parse(resourceObject: Any?, throwIfWrongType: Boolean): ResourcesConfig? {
        if (resourceObject is JSONObject) {
            return parseFromJson(resourceObject, throwIfWrongType)
        } else if (throwIfWrongType) {
            throw InvalidConfigException("Source must be a JSONObject or JSONArray.")
        }
        return null
    }

    @Throws(InvalidConfigException::class)
    protected fun parseFromJson(configJson: JSONObject, throwIfWrongType: Boolean): ResourcesConfig? {
        val type = JsonParseUtils.getString(configJson, TYPE, SOURCE, false)
        val builder = builderByType(type, throwIfWrongType) ?: return null
        fillFromJSON(builder, configJson)
        validate(builder)
        return builder.build()
    }

    /**
     * Fill source object from JSON.
     *
     * @param builder      Source to fill.
     * @param configJson JSON object that contain source config properties.
     * @throws InvalidConfigException if config has some logic errors or doesn't contain some required fields.
     */
    @Throws(InvalidConfigException::class)
    protected open fun fillFromJSON(builder: ConfigBuilder, configJson: JSONObject) {
        builder.keyColumn = JsonParseUtils.getString(configJson, COLUMN_KEY, SOURCE, true)
            ?: throw InvalidConfigException("\"$SOURCE.$COLUMN_KEY\" is not set.")
        builder.quantityColumn = JsonParseUtils.getString(configJson, COLUMN_QUANTITY, SOURCE, false)
        builder.commentColumn = JsonParseUtils.getString(configJson, COLUMN_COMMENT, SOURCE, false)
        builder.metadataColumn = JsonParseUtils.getString(configJson, COLUMN_METADATA, SOURCE, false)
    }

    @Throws(InvalidConfigException::class)
    protected open fun validate(builder: ConfigBuilder) {
        if (builder.keyColumn.isEmpty())
            throw InvalidConfigException("\"$SOURCE.$COLUMN_KEY\" is not set.")
    }
}